x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -2324,6 +2325,18 @@ public ApiKeyDoc toApiKeyDoc(BytesReference roleDescriptorsBytes, BytesReference
         }
     }
 
+    /**
+     * API Key documents are refreshed after creation, such that the API Key docs are visible in searches after the create-API-key
+     * endpoint returns.
+     * In stateful deployments, the automatic refresh interval is short (hard-coded to 1 sec), so the {@code RefreshPolicy#WAIT_UNTIL}
+     * is an acceptable tradeoff for the superior doc creation throughput compared to {@code RefreshPolicy#IMMEDIATE}.
+     * But in stateless the automatic refresh interval is too long (at least 10 sec), which translates to long create-API-key endpoint
+     * latency, so in this case we opt for {@code RefreshPolicy#IMMEDIATE} and acknowledge the lower maximum doc creation throughput.
+     */
+    public static RefreshPolicy defaultCreateDocRefreshPolicy(Settings settings) {
+        return DiscoveryNode.isStateless(settings) ? RefreshPolicy.IMMEDIATE : RefreshPolicy.WAIT_UNTIL;
```

Comment:
This works, but ideally this would be based on the actual refresh interval for the .security index instead of the code assuming the refresh interval for stateless (which is true today, but maybe not tomorrow). Are there any reasonable path's to determine the refresh interval ? Ideally we could just read it from the settings, but not sure how default index refresh intervals are modeled.

Reply:
Since the security index depends on the default refresh interval, we can just query that statically instead of using the actual index settings. The default refresh interval is defined [here](https://github.com/elastic/elasticsearch/blob/bd4dcbd4a8855544c973f5556953cc4b817598b3/server/src/main/java/org/elasticsearch/index/IndexSettings.java#L286-L293)

We can do something similar to the follows:
```java
final var refreshInternal = INDEX_REFRESH_INTERVAL_SETTING.get(Settings.EMPTY);
...
```

I personally don't mind just using `isStateless()` here. What we want is to immediately refresh if the auto-refresh is slow. But slow is subjective and we need to come up with a concrete number. Is it 10 seconds, 5 seconds or 2 seconds? That seems additionally complexity. That said, I don't feel strongly if we choose to branch on refresh interval value.

Reply:
> Are there any reasonable path's to determine the refresh interval ? Ideally we could just read it from the settings, but not sure how default index refresh intervals are modeled.

We could use the default index refresh setting, as Yang is describing.
But I don't think that's an improvement. Fundamentally stateless refresh take longer and are more expensive so they are rarer. It seems brittle to put a cutoff on "rarer" specifically for deciding the API Key doc indexing refresh policy. Because we can't change the refresh interval for system indices makes it less brittle in practice, but that's a weak assumption to rely on, I think.

Using the dynamic refresh interval value (so not relying on the immutability of the automatic refresh interval for system indices) adds runtime overhead and complicates the code (caching some recent value).

Reply:
Fair enough, i still feel the decision should be a function of interval and not the deployment but it practice it likely does not matter.